src/common/buffer.cc

```diff
@@ -1419,28 +1419,26 @@ using namespace ceph;
   void buffer::list::rebuild()
   {
     if (_len == 0) {
-      _buffers.clear();
+      _buffers.clear_and_dispose(hangable_ptr::disposer());
       return;
     }
-    ptr nb;
     if ((_len & ~CEPH_PAGE_MASK) == 0)
-      nb = buffer::create_page_aligned(_len);
+      rebuild(hangable_ptr::create(buffer::create_page_aligned(_len)));
     else
-      nb = buffer::create(_len);
-    rebuild(nb);
+      rebuild(hangable_ptr::create(buffer::create(_len)));
   }
 
-  void buffer::list::rebuild(ptr& nb)
+  void buffer::list::rebuild(hangable_ptr& nb)
   {
     unsigned pos = 0;
-    for (std::list<ptr>::iterator it = _buffers.begin();
+    for (buffers_t::iterator it = _buffers.begin();
 	 it != _buffers.end();
 	 ++it) {
       nb.copy_in(pos, it->length(), it->c_str(), false);
       pos += it->length();
     }
     _memcopy_count += pos;
-    _buffers.clear();
+    _buffers.clear_and_dispose(hangable_ptr::disposer());
```

Comment:
This is really good catch. Thanks, @cbodley!

Reply:
For intermediary commits introduced the manual disposing:
```cpp
    } else {
      ptr_node::disposer()(&nb);
    }
```

Additionally pushed a new that switches all `ptr_node` factories to `std::unique_ptr<..., disposer>` to prevent such issues in the future.